lib/core/async/event_synchronizer.dart

```diff
@@ -0,0 +1,145 @@
+import 'dart:async';
+
+import 'package:uuid/uuid.dart';
+
+/// A function which dispatches messages over I/O ports.
+typedef Sender = Future<void> Function(dynamic);
+
+/// A subscriber is a function that receives messages from [handleMessage] and acts on them.
+///
+/// A subscriber is short lived and is killed after the first call.
+/// An exception is when the subscriber returns `true`. In this case
+/// the subsciber is kept alive and will be called on upcoming messages until
+/// it returns `null` or `false`.
+typedef _Subscriber = Future<bool?> Function(dynamic);
+
+/// A class which helps managing asynchronuous, callback-based communnication.
+///
+/// [EventSynchronizer] allows to request, expect and exchange messages between two independent
+/// endpoints, e.g. web sockets, isolates, ...
+///
+/// ```dart
+/// // Some I/O client which sends and receives messages.
+/// final client = IOCLient()
+///
+/// // The synchronizer is initialized with the send function of the client.
+/// final synchronizer = EventSynchronizer(client.send);
+///
+/// // We need to pass each message from the client to the synchronizer.
+/// client.onMessage = (dynamic message) => synchronizer.handleMessage(message);
+///
+/// // Simple messages are sent using the request function.
+/// await synchronizer.request('ANYTHING');
+///
+/// // We can receive the String response directly using an identity function.
+/// final String response = await synchronizer.expect(String, (String resp) => resp);
+///
+/// // If we are witing for a specific response and can ignore other messages, we can use keepAlive.
+/// // It will wait with completion until we receive "THAT'S IT".
+/// // The return value in this case is `true`.
+/// await synchronizer.expect(String, (String resp) => resp == "THAT'S IT", keepAlive: true);
+///
+/// // Using exhange, we can reallize a ping-pong communication.
+/// // Use drain to await the stream completion.
+/// synchronizer.exchange(someStream, String, (String req, String resp) => resp).listen((String resp) {
+///  print(resp);
+/// });
+/// ```
+class EventSynchronizer {
+  /// The sener function used to dispatch messages.
+  final Sender _sender;
+
+  /// A map of message subscriptions identified by a random UUID.
+  final Map<String, _Subscriber> _subscriptions = {};
+
+  /// UUID generator.
+  final _uuid = Uuid();
+
+  /// A time offset required if the I/O responses might be faster than the code which attaches
+  /// a listener ([expect]) for them.
+  ///
+  /// The duration is a minimum time which is required for each [handleMessage] call.
+  /// Use with caution!
+  final Duration messageOffset;
+
+  EventSynchronizer(this._sender, [this.messageOffset = Duration.zero]);
+
+  /// Message handler which should be called for each incoming [message] without pre-filtering.
+  Future<void> handleMessage(dynamic message) async {
```

Comment:
Because it's part of the public API I don't want to assume that there's not a single application where it might be required to await the completion of `handleMessage`.

Reply:
Fair enough